crates/transpiler/src/equivalence.rs

```diff
@@ -590,16 +590,15 @@ impl EquivalenceLibrary {
     /// (including those from base).
     pub fn add_equivalence(
         &mut self,
-        py: Python,
         gate: &PackedOperation,
         params: &[Param],
-        equivalent_circuit: CircuitFromPython,
+        equivalent_circuit: CircuitData,
     ) -> PyResult<()> {
         raise_if_shape_mismatch(gate, &equivalent_circuit)?;
-        raise_if_param_mismatch(py, params, equivalent_circuit.0.unsorted_parameters(py)?)?;
+        raise_if_param_mismatch(params, equivalent_circuit.parameters())?;
         let key: Key = Key::from_operation(gate);
         let equiv = Equivalence {
-            circuit: equivalent_circuit.clone(),
+            circuit: CircuitFromPython(equivalent_circuit.clone()),
```

Comment:
Why do you need to use the newtype here? It seems a bit odd that this is needed for anything anymore once we're in a Rust space function?

Like I assume it's because you have `IntoPyObject` implemented for the newtype and you're using that as the return type from a a pymethod. But I think a stricter separation where you just normalize on using `CircuitData` insde the rust domain makes the most sense. You can do the conversion to a Python space `QuantumCircuit` in the python interface layer if you're returning `Equivalence` anywhere directly.

Reply:
FIxed in 10cf87a